python/ql/test/experimental/dataflow/tainttracking/customSanitizer/TestTaint.expected

```diff
@@ -1,9 +1,25 @@
 test_taint
 | test.py:22 | ok   | test_custom_sanitizer | s |
-| test.py:36 | fail | test_custom_sanitizer_guard | s |
+| test.py:36 | ok   | test_custom_sanitizer_guard | s |
 | test.py:38 | ok   | test_custom_sanitizer_guard | s |
-| test.py:49 | ok   | test_escape | s2 |
+| test.py:40 | ok   | test_custom_sanitizer_guard | s |
+| test.py:51 | ok   | test_escape | s2 |
 isSanitizer
 | TestTaintTrackingConfiguration | test.py:21:39:21:39 | ControlFlowNode for s |
-| TestTaintTrackingConfiguration | test.py:48:10:48:29 | ControlFlowNode for emulated_escaping() |
+| TestTaintTrackingConfiguration | test.py:50:10:50:29 | ControlFlowNode for emulated_escaping() |
 isSanitizerGuard
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() |
+sanitizerGuardControls
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:36:9:36:26 | ControlFlowNode for ensure_not_tainted | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:36:9:36:29 | ControlFlowNode for ensure_not_tainted() | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:36:28:36:28 | ControlFlowNode for s | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:37:9:37:9 | ControlFlowNode for s | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:37:13:37:26 | ControlFlowNode for TAINTED_STRING | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:38:9:38:22 | ControlFlowNode for ensure_tainted | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:38:9:38:25 | ControlFlowNode for ensure_tainted() | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:38:24:38:24 | ControlFlowNode for s | true |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:40:9:40:22 | ControlFlowNode for ensure_tainted | false |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:40:9:40:25 | ControlFlowNode for ensure_tainted() | false |
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:40:24:40:24 | ControlFlowNode for s | false |
+sanitizerGuardedNode
+| TestTaintTrackingConfiguration | test.py:35:8:35:26 | ControlFlowNode for emulated_is_safe() | test.py:36:28:36:28 | ControlFlowNode for s |
```

Comment:
This output is quite verbose, and I guess it has been helpful for you to see that the internal details of guards and what they control. Can we move this part somewhere else more appropriate?

I'm not feeling SUPER strong about this one (I'm probably at 4/10), but I think it would be really nice to not leave the custom sanitizer with this debt of also needing to check on the internal implementation.

Reply:
I didn't really think too much until now, sorry :disappointed:

Reply:
Would you remove both `sanitizerGuardControls` and `sanitizerGuardedNode`?
I would actually like a strategy for what might be considered "debug queries", there are a few in the basic dataflow test already. Should we
- have them run on very small test files?
- have them commented out somewhere?
- reconstruct them every time we debug?
- have them run when "explicitly turned on" (is that possible? via some flag or a specific debug test suite, perhaps)?

As you say, they can be quite verbose..

Reply:
> Would you remove both sanitizerGuardControls and sanitizerGuardedNode?

yes

### strategy for debug queries

_reconstruct them every time we debug_ doesn't sound very good. How about adding them as query predicates to a qll file that can be imported if you need to debug something? (having it in a ql file would not be good enough, since configuration needs to be in scope as well) I guess we could place this in `src/experimental/meta/debug/dataflow.qll` or similar. That would also make it easy to run against snapshots when we need to :wink:

Maybe that also becomes a bit too over engineered, and simply having them all in a test-file where you can copy-paste them from is good enough. 

Leaving them commented out doesn't sound like a good plan to me. Actually, having them executed as part of test-suite sounds really good.

Reply:
I have gone for the debug library option. It may be over-engineered, but I like how it creates a place to shunt everything that has been useful at some point and not entirely trivial to create. I can sense that it will encourage me to put many things there, leaving the normal test output cleaner.